docs/guide/api-plugin.md

```diff
@@ -423,10 +424,31 @@ Vite plugins can also provide hooks that serve Vite-specific purposes. These hoo
 
   - Filter and narrow down the affected module list so that the HMR is more accurate.
 
-  - Return an empty array and perform complete custom HMR handling by sending custom events to the client (example uses `server.hot` which was introduced in Vite 5.1, it is recommended to also use `server.ws` if you support lower versions):
+  - Return an empty array and perform a full reload:
+
+    ```js
+    handleHotUpdate({ server, modules, timestamp }) {
+      // Also use `server.ws.send` to support Vite <5.1 if needed
+      server.hot.send({ type: 'full-reload' })
+      // Invalidate modules manually
+      const invalidatedModules = new Set()
+      for (const mod of modules) {
+        server.moduleGraph.invalidateModule(
+          mod,
+          invalidatedModules,
+          timestamp,
+          true
+        )
+      }
```

Comment:
Ideally I think, we shouldn't need to have user invalidate this way. This line

https://github.com/vitejs/vite/blob/e4ea7dd478765665137c23aeac3a8a888b952187/packages/vite/src/node/server/index.ts#L723-L724

Should already cover the invalidation, but this code here is to also invalidate for HMR (forth param). The param is used for 

https://github.com/vitejs/vite/blob/e4ea7dd478765665137c23aeac3a8a888b952187/packages/vite/src/node/server/moduleGraph.ts#L186-L192

But I was thinking of combining the two timestamps in the future. And then we can remove the HMR param.